### PR TITLE
Add NoUselessElseFixer

### PR DESCRIPTION
**Pull request by SpacePossum**
todo:
integration tests

replaces:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1693

### CODE REVIEW

**Comment by keradus** [comment]
Also, Travis is red

**Comment by SpacePossum** [comment]
that is because f the TODO, the integration tests are not done, just stubbed :)

**Comment by keradus** [comment]
OK, cool ;)

File: Symfony/CS/Fixer/Contrib/NoUselessElseFixer.php [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+                return;
+            }
+        } while (!$tokens[$previousBlockStart]->isGivenKind(T_IF));
+
+        // if we made it to here the 'else' can be removed
+        $this->clearElse($tokens, $index);
+    }
+
+    /**
+     * Return the first and lost token index of the previous block.
```
Comment (gharlan) [inline Symfony/CS/Fixer/Contrib/NoUselessElseFixer.php]: "last"

**Comment by gharlan** [comment]
Is there a test case similar to this?

```php
if ($x) {
    if ($y) {
        return 1;
    } else {
        return 2;
    }
} else {
    return 3;
}
```

**Comment by SpacePossum** [comment]
TODO:
HHVM 3.6.1 https://3v4l.org/U2hoG tokenizes diff. than 3.8.1 / PHP Zend

**Comment by keradus** [comment]
Ref https://github.com/keradus/PHP-CS-Fixer/commit/c6761cd74c6a8ac611936b9abdab6785217674a2

**Comment by SpacePossum** [comment]
Should stay @ contrib level. I think this one is ready, please play around with it :)

File: Symfony/CS/Fixer/Contrib/NoUselessElseFixer.php [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function fix(\SplFileInfo $file, $content)
+    {
+        $tokens = Tokens::fromCode($content);
+
+        // start from 4; minimum number of tokens that form a candidate for fixing; <?php if ( [] ) [5]
+        foreach ($tokens as $index => $token) {
```
Comment (keradus) [inline Symfony/CS/Fixer/Contrib/NoUselessElseFixer.php]: either fix the loop to respect the comment or remove the misleading comment

**Comment by keradus** [comment]
Could anyone handle the review?

**Comment by gharlan** [comment]
https://github.com/SpacePossum/PHP-CS-Fixer/pull/11

**Comment by keradus** [comment]
@SpacePossum could you squash commits of yours into one and then leave @gharlan at the end?

**Comment by SpacePossum** [comment]
>@SpacePossum could you squash commits of yours into one and then leave @gharlan at the end?

Sure thing :)
I'll work on this a bit more, little refactor, and than squash after the review.

**Comment by keradus** [comment]
Oh, I thought it's ready to merge.
When done please remove WIP label

**Comment by SpacePossum** [comment]
Happy to take review comments and thoughts on this one to wrap up it up :)
ping @gharlan and @keradus and everyone else.

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 3]
```diff
@@ -180,7 +180,8 @@ protected function doTestIntegration(IntegrationCase $case)
             throw new IOException(sprintf('Failed to write to tmp. file "%s".', $tmpFile));
         }
 
-        $changed = $fixer->fixFile(new \SplFileInfo($tmpFile), $case->getFixers(), false, true, new FileCacheManager(false, null, $case->getFixers()));
+        $fixers = $case->getFixers();
+        $changed = $fixer->fixFile(new \SplFileInfo($tmpFile), $case->getFixers(), false, true, new FileCacheManager(false, null, $fixers));
```
Comment (keradus) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: $case->getFixers() => $fixers

**Comment by SpacePossum** [comment]
updated and squashed

File: Symfony/CS/Fixer/Contrib/NoUselessElseFixer.php [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+namespace Symfony\CS\Fixer\Contrib;
+
+use Symfony\CS\AbstractAlignFixer;
+use Symfony\CS\Tokenizer\Tokens;
+
+/**
+ * @author SpacePossum
+ */
+final class NoUselessElseFixer extends AbstractAlignFixer
```
Comment (keradus) [inline Symfony/CS/Fixer/Contrib/NoUselessElseFixer.php]:
```diff
-AbstractAlignFixer
+AbstractFixer
```

Comment (SpacePossum) [inline Symfony/CS/Fixer/Contrib/NoUselessElseFixer.php]: :+1:  done

**Comment by GrahamCampbell** [comment]
Missing tests for else with colons?

**Comment by SpacePossum** [comment]
wish they killed the alternative in PHP7... oh well.. lets see...

**Comment by GrahamCampbell** [comment]
#PHP8 :P

**Comment by SpacePossum** [comment]
>Missing tests for else with colons?

Test and a minor speed gain added. I didn't put effort in detecting / cleanup useless else statements done in the alternative syntax because it is no fun and an edge case anyway.

File: Symfony/CS/Tests/AbstractIntegrationTest.php [hunk 5]
```diff
@@ -180,7 +180,8 @@ protected function doTestIntegration(IntegrationCase $case)
             throw new IOException(sprintf('Failed to write to tmp. file "%s".', $tmpFile));
         }
 
-        $changed = $fixer->fixFile(new \SplFileInfo($tmpFile), $case->getFixers(), false, true, new FileCacheManager(false, null, $case->getFixers()));
+        $fixers = $case->getFixers();
+        $changed = $fixer->fixFile(new \SplFileInfo($tmpFile), $fixers, false, true, new FileCacheManager(false, null, $fixers));
```
Comment (gharlan) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: Unrelated to this pr, right?

Comment (SpacePossum) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: indeed, however I think it is of value, should I take it out?

Comment (gharlan) [inline Symfony/CS/Tests/AbstractIntegrationTest.php]: In my opinion it should be another pr, together with the change in `IntegrationCase`.

**Comment by SpacePossum** [comment]
Updated, rebased and clean upped.
The commit history looks a bit odd now with @gharlan being the author of the whole test file, hope he doesn't mind.

**Comment by keradus** [comment]
Please, could you recommit to retrigger Travis? Also, put no_useless_else to .php_cs file

**Comment by GrahamCampbell** [comment]
@keradus Or just close and re-open the PR.

**Comment by keradus** [comment]
Didn't know that trick, thanks ;)
(But it's sad we need to have tricks for Travis :/)

**Comment by gharlan** [comment]
@keradus you should be able to restart builds because you have write permissions to the repo.

screenshot 2016-05-23 21 45 41

**Comment by keradus** [comment]
@gharlan usually it is like that and I do it often, sadly this time it wasn't about crashed build, but the build was not triggered at all, so I have no build to restart  - I needed to retrigger it ;)

**Comment by keradus** [comment]
Thank you @SpacePossum.

**Comment by keradus** [comment]
Thank you @gharlan

**Comment by SpacePossum** [comment]
: D
thanks @gharlan for additional test and reviews, @keradus and @GrahamCampbell thanks for reviewing  :+1:

**Comment by localheinz** [comment]
@SpacePossum 

👍 Great fixer, but you wouldn't believe the code that was touched by it - I'm looking at it right now, and my eyes are bleeding! Nice job!
